lib/queuePopulator/LogReader.js

```diff
@@ -712,25 +712,50 @@ class LogReader {
         });
     }
 
+    /**
+     * Determine if the log offset should be updated.
+     * @param {Object} logRes - Log response object
+     * @param {Number} nextLogOffset - Next log offset
+     * @returns {Boolean} - True if the log offset should be updated
+     */
+    _shouldUpdateLogOffset(logRes, nextLogOffset) {
+        return nextLogOffset !== undefined &&
+            nextLogOffset !== this.logOffset &&
+            (!logRes.log.reachedUnpublishedListing || logRes.log.reachedUnpublishedListing());
+    }
+
+    /**
+     * Update metrics for Raft backend.
+     * @param {Object} logInfo - Log information object
+     * @returns {undefined}
+     */
+    _updateMetrics(logInfo) {
+        const logSize = logInfo.cseq || logInfo.end;
+        this._metricsHandler.logReadOffset(this.getMetricLabels(), this.logOffset);
+        this._metricsHandler.logSize(this.getMetricLabels(), logSize);
+    }
+    
     _processSaveLogOffset(batchState, done) {
         const { logRes, nextLogOffset, logger } = batchState;
-        if (this.isOffsetManaged() &&
-            nextLogOffset !== undefined &&
-            nextLogOffset !== this.logOffset &&
-            (!logRes.log.reachedUnpublishedListing ||
-             logRes.log.reachedUnpublishedListing())) {
 
+        // Handle offset managed by the log reader (stored in Zookeeper)
+        if (this.isOffsetManaged() && this._shouldUpdateLogOffset(logRes, nextLogOffset)) {
             this.logOffset = nextLogOffset;
 
-            // Skip metrics if we are not using Raft backend
+            // Update metrics for Raft backend
             if (logRes.info) {
-                const logSize = logRes.info.cseq || logRes.info.end;
-                this._metricsHandler.logReadOffset(this.getMetricLabels(), this.logOffset);
-                this._metricsHandler.logSize(this.getMetricLabels(), logSize);
+                this._updateMetrics(logRes.info);
             }
 
             return this._writeLogOffset(logger, done);
         }
+
+        // Handle offset managed externally (e.g., Kafka)
+        if (!this.isOffsetManaged() && this.logConsumer.storeOffsets) {
```

Comment:
When `!this.isOffsetManaged()`, then `storeOffsets` should always be defined, right?
If this is the case, then this option is redundant (may be an assertion in constructor)